keps/sig-storage/20200120-skip-permission-change.md

```diff
@@ -0,0 +1,134 @@
+---
+title: Skip Volume Ownership Change
+authors:
+  - "@gnuified"
+owning-sig: sig-storage
+participating-sigs: sig-auth
+reviewers:
+  - "@msau42"
+  - "@liggit"
+  - "@tallclair"
+approvers:
+  - "@saad-ali"
+editor: TBD
+creation-date: 2020-01-20
+last-updated: 2020-01-20
+status: implementable
+see-also:
+replaces:
+superseded-by:
+
+---
+
+# Skip Volume Ownership Change
+
+## Table of Contents
+
+* [Table of Contents](#table-of-contents)
+* [Summary](#summary)
+* [Motivation](#motivation)
+    * [Goals](#goals)
+    * [Non-Goals](#non-goals)
+* [Proposal](#proposal)
+    * [Implementation Details/Notes/Constraints [optional]](#implementation-detailsnotesconstraints-optional)
+    * [Risks and Mitigations](#risks-and-mitigations)
+* [Graduation Criteria](#graduation-criteria)
+* [Implementation History](#implementation-history)
+* [Drawbacks [optional]](#drawbacks-optional)
+* [Alternatives [optional]](#alternatives-optional)
+
+## Summary
+
+Currently before a volume is bind-mounted inside a container the permissions on
+that volume are changed recursively to the provided fsGroup value.  This change
+in ownership can take an excessively long time to complete, especially for very
+large volumes (>=1TB) as well as a few other reasons detailed in [Motivation].
+To solve this issue we will add a new field called `VolumePermissionChangePolicy` and
+allow the user to specify whether they want the ownership change to occur.
+
+## Motivation
+
+When a volume is mounted on the node, we recursively change permissions of volume
+before bind mounting the volume inside container. The reason of doing this is to ensure
+that volumes are readable/writable by provided fsGroup.
+
+But this presents following problems:
+ - An application(many popular databases) which is sensitive to permission bits changing
+   underneath may refuse to start whenever volume being used inside pod gets mounted on
+   different node.
+ - If volume has a large number of files, performing recursive `chown` and `chmod`
+   could be slow and could cause timeout while starting the pod.
+
+### Goals
+
+ - Allow volume ownership and permission to be skipped during mount
+
+### Non-Goals
+
+ - In some cases if user brings in a large enough volume from outside, the first time ownership and permission change still could take lot of time.
+ - On SELinux enabled distributions we will still do recursive chcon whenever applicable and handling that is outside the scope.
+
+## Proposal
+
+We propose that an user can optionally opt-in to skip recursive ownership(and permission) change on the volume if volume already has right permissions.
+
+### Implementation Details/Notes/Constraints [optional]
+
+When creating a pod, we propose that `pod.Spec.SecurityContext` field expanded to include a new field called `VolumePermissionChangePolicy` which can have following possible values:
+
+ - `NoChange` --> Don't change permissions and ownership.
```

Comment:
How is it different from not having fsGroup at all?

Reply:
I am thinking that in k8s deployments where user can't explicitly out-out of having no fsGroup, this could be a way out for an user. IIRC - in Openshift a non-privileged user who only has restricted SCC can't opt-out of having fsGroup (I could verify this).

Reply:
Please include this reason in the KEP.